ipa-core/src/protocol/ipa_prf/prf_sharding/feature_label_dot_product.rs

```diff
@@ -213,90 +212,72 @@ where
-    let prime_field_validator = sh_ctx.narrow(&Step::PrimeFieldValidator).validator::<F>();
-    let prime_field_ctx = prime_field_validator.context();
-
     // Tricky hacks to work around the limitations of our current infrastructure
     let num_outputs = input_rows.len() - histogram[0];
-    let mut record_id_for_row_depth = vec![0_u32; histogram.len()];
     let ctx_for_row_number = set_up_contexts(&binary_m_ctx, histogram);
 
     // Chunk the incoming stream of records into stream of vectors of records with the same PRF
-    let mut input_stream = stream_iter(input_rows);
-    let first_row = input_stream.next().await;
-    if first_row.is_none() {
+    let mut input_stream = stream::iter(input_rows);
+    let Some(first_row) = input_stream.next().await else {
         return Ok(vec![]);
-    }
-    let first_row = first_row.unwrap();
+    };
     let rows_chunked_by_user = chunk_rows_by_user(input_stream, first_row);
 
-    // Convert to a stream of async futures that represent the result of executing the per-user circuit
-    let stream_of_per_user_circuits = pin!(rows_chunked_by_user.then(|rows_for_user| {
-        let num_user_rows = rows_for_user.len();
-        let contexts = ctx_for_row_number[..num_user_rows - 1].to_owned();
-        let record_ids = record_id_for_row_depth[..num_user_rows].to_owned();
-        record_id_for_row_depth[..num_user_rows]
-            .iter_mut()
-            .for_each(|count| *count += 1);
-
-        #[allow(clippy::async_yields_async)]
-        // this is ok, because seq join wants a stream of futures
-        async move {
-            evaluate_per_user_attribution_circuit(contexts, record_ids, rows_for_user)
-        }
-    }));
+    let mut collected = rows_chunked_by_user.collect::<Vec<_>>().await;
+    collected.sort_by(|a, b| std::cmp::Ord::cmp(&b.len(), &a.len()));
+
+    let chunked_user_results =
+        collected
+            .into_iter()
+            .enumerate()
+            .map(|(record_id, rows_for_user)| {
+                let num_user_rows = rows_for_user.len();
+                let contexts = ctx_for_row_number[..num_user_rows - 1].to_owned();
+
+                evaluate_per_user_attribution_circuit(
+                    contexts,
+                    RecordId::from(record_id),
+                    rows_for_user,
+                )
+            });
 
     // Execute all of the async futures (sequentially), and flatten the result
-    let flattened_stream = seq_join(sh_ctx.active_work(), stream_of_per_user_circuits)
-        .flat_map(|x| stream_iter(x.unwrap()));
-
-    // modulus convert feature vector bits from shares in `Z_2` to shares in `Z_p`
-    let converted_feature_vector_bits = convert_bits(
-        prime_field_ctx
-            .narrow(&Step::ModulusConvertFeatureVectorBits)
-            .set_total_records(num_outputs),
-        flattened_stream,
-        0..<FV as SharedValue>::BITS,
+    let flattened_stream = Box::pin(
```

Comment:
does this now work w/o multithreading or we still rely on it being on for aggregation?

Reply:
I believe it is still required, perhaps even more so. We can add the buffered stream adapter if we really want things to work without multi-threading.